### PR TITLE
fix check before creating a table

### LINKED ISSUE #8091: table creation shouldn't be allowed if table delete is happening async in background
Currently table is deleted first and the segments were deleted async behind the scene. 

however if the controller died during the async deletion and got restarted. User can call the API to create an exact same name table without any issue. This should not be allowed since there's still left-over segments. 

I am not sure what's the right way to fix it. but some thoughts:
1. create a soft delete state for a table marking it as deleting, and not allow same name table to be recreated until the segment deletion is done. 
2. check the property store as well as table config store to make sure no left over segments exist, otherwise reject the table config creation request. 

Thoughts?

### PR DESCRIPTION
**Pull request by walterddr**
Summary
===
This fixes #8091. table creation should check that table config doesn't exist.

Release Note
===
REALTIME table now doesn't support creation of a duplicate table name. It will now fail with a `TableAlreadyExistException`

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 1]
```diff
@@ -1270,6 +1270,11 @@ public void addTable(TableConfig tableConfig)
         if (getAllTables().contains(tableNameWithType)) {
           throw new TableAlreadyExistsException("Table " + tableNameWithType + " already exists");
         }
+        List<SegmentZKMetadata> segmentsZKMetadata = getSegmentsZKMetadata(tableNameWithType);
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:1273]: No need to actually read the `SegmentZKMetadata`, checking if the dir exists should be enough

**Review by Jackie-Jiang — COMMENTED** [review]
Per the discussion in #8091, we should also check other fields, e.g. linage etc.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java (hunk 1 above)
Comment (walterddr) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:1273]: should we at least list out the dir content? (would there ever be an empty dir?)

Comment (mcvsubbu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:1273]: If you list things out, please limit the list size to (say) 5 segments. You can indicate in the message that there are more (e.g. `...` could be used).

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 2]
```diff
@@ -1270,6 +1270,11 @@ public void addTable(TableConfig tableConfig)
...
+        if (segmentsZKMetadata != null && segmentsZKMetadata.size() != 0) {
+          throw new TableAlreadyExistsException("Table " + tableNameWithType + " still have active segments registered"
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:1275]: Table already exists may not be the right one, depending on whether the table is under deletion now? Is it guaranteed that the delete call removes all metadata right away?

Comment (walterddr) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:1275]: nope. it doesn't, however I can't find a better exception type. should I create a new one?

Comment (mcvsubbu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:1275]: you can change the exception message accordingly.  Provide as accurate a message as possible

**Comment by walterddr** [comment]
> Per the discussion in #8091, we should also check other fields, e.g. linage etc.

added segment lineage check, partition metadata seems to be in the segment ZNRecord so we are good; i am not sure what's the task metadata we should check. please let me know if this looks good.

**Comment by Jackie-Jiang** [comment]
I think there is an easier way to make sure everything is cleaned up:
- In `deleteOfflineTable()` and `deleteRealtimeTable()`, move the step of removing table config as the last step
- When creating a new table, simply checking table config existence should be enough

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 3]
```diff
@@ -1261,13 +1261,12 @@ public void addTable(TableConfig tableConfig)
     validateTableTenantConfig(tableConfig);
     String tableNameWithType = tableConfig.getTableName();
     SegmentsValidationAndRetentionConfig segmentsConfig = tableConfig.getValidationConfig();
-
     TableType tableType = tableConfig.getTableType();
     switch (tableType) {
       case OFFLINE:
         // existing tooling relies on this check not existing for realtime table (to migrate to LLC)
```
Comment (walterddr) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:1267]: is this issue fixed? we should probably do this check on all table types.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 4]
```diff
@@ -1262,14 +1262,12 @@ public void addTable(TableConfig tableConfig)
     String tableNameWithType = tableConfig.getTableName();
     SegmentsValidationAndRetentionConfig segmentsConfig = tableConfig.getValidationConfig();
 
+    if (getTableConfig(tableNameWithType) != null) {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:1265]: (minor) Suggest moving this as the first step of table creation

**Review by Jackie-Jiang — APPROVED** [review]
LGTM

@siddharthteotia Can you please take a look and see if adding this check to RT table is okay. Per the comment, the old behavior is for HLC to LLC migration tooling, so I feel it is okay to make both OFFLINE and REALTIME table's behavior consistent

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java (hunk 4 above)
Comment (walterddr) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:1265]: will wait for CI & @siddharthteotia's comment before making this final adjustment

**Pull request merged**